### PR TITLE
Improve parallelCoordinates

### PR DESCRIPTION
**Pull request by mgalloue**
This is a beginning of a series of improvements: 

- Highlight a line on mouse over
- Possibility to change lines tension
- Refactoring
- Support axis drag and drop

### CODE REVIEW

**Comment by liquidpele** [comment]
Your PR is removing/undoing unrelated stuff....

File: src/css/parallelcoordinates.css [hunk 1]
```diff
@@ -25,3 +25,8 @@
 .nvd3 .axis text {
     text-shadow: 0 1px 0 #fff;
 }
+
+.nvd3 .hover  {
+    fill-opacity: 1;
+	stroke-width: 3px;
+}
```
Comment (liquidpele) [inline src/css/parallelcoordinates.css]: This will apply to way more than parallel coordinates, you need to have a class before .hover that's specific to this chart.

File: src/models/parallelCoordinates.js [hunk 2]
```diff
@@ -158,13 +170,14 @@ nv.models.parallelCoordinates = function() {
...
         // options that require extra logic in the setter
         margin: {get: function(){return margin;}, set: function(_){
-            margin.top    = typeof _.top    != 'undefined' ? _.top    : margin.top;
-            margin.right  = typeof _.right  != 'undefined' ? _.right  : margin.right;
-            margin.bottom = typeof _.bottom != 'undefined' ? _.bottom : margin.bottom;
-            margin.left   = typeof _.left   != 'undefined' ? _.left   : margin.left;
+            margin.top    =  _.top    != undefined ? _.top    : margin.top;
+            margin.right  =  _.right  != undefined ? _.right  : margin.right;
+            margin.bottom =  _.bottom != undefined ? _.bottom : margin.bottom;
+            margin.left   =  _.left   != undefined ? _.left   : margin.left;
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: You need to do !== instead of != to ensure it's actually undefined, otherwise you can't ever set it to 0.

File: src/models/parallelCoordinates.js [hunk 3]
```diff
@@ -9,16 +9,19 @@ nv.models.parallelCoordinates = function() {
...
+        , height = 500
         , x = d3.scale.ordinal()
         , y = {}
         , dimensions = []
         , color = nv.utils.defaultColor()
         , filters = []
         , active = []
-        , dispatch = d3.dispatch('brush')
+        , foreground
+        , background
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: foreground and background are only used inside the chart(){...} block, so just make them vars inside there.   The block of vars up here is meant to be for persistent settings.

File: src/models/parallelCoordinates.js [hunk 4]
```diff
@@ -58,32 +58,44 @@ nv.models.parallelCoordinates = function() {
...
+            gEnter.append('g').attr('class', 'foreground');
+
             wrap.attr('transform', 'translate(' + margin.left + ',' + margin.top + ')');
 
-            var line = d3.svg.line(),
-                axis = d3.svg.axis().orient('left'),
-                background,
-                foreground;
+            var line = d3.svg.line().interpolate('cardinal').tension(lineTension),
+                axis = d3.svg.axis().orient('left');
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: axis doesn't appear to be ever used?

Comment (mgalloue) [inline src/models/parallelCoordinates.js]: It's used line 110

 dimension.append('g')
                .attr('class', 'axis')
                .each(function(d) { d3.select(this).call(axis.scale(y[d])); })

Comment (liquidpele) [inline src/models/parallelCoordinates.js]: ah okay, didn't see that :)

File: src/models/parallelCoordinates.js [hunk 5]
```diff
@@ -9,16 +9,17 @@ nv.models.parallelCoordinates = function() {
     // Public Variables with Default Settings
     //------------------------------------------------------------
 
-    var margin = {top: 30, right: 10, bottom: 10, left: 10}
-        , width = null
-        , height = null
+    var margin = {top: 30, right: 0, bottom: 10, left: 0}
+        , width = 1000
+        , height = 500
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: These should start off being null so it uses the full width and height of the SVG, if you need specific values the chart.width() and chart.height() options should be used.

**Comment by liquidpele** [comment]
It looks like there is a conflict since github can't automatically merge... you shoudl be able to fix that by creating the upstream repo like so:

https://help.github.com/articles/configuring-a-remote-for-a-fork/

Then just do "git fetch upstream" and "git merge upstream/development" into your own branch to sync it up and fix any conflicts.  Commit your conflict resolution and then push it up and it shoudl fix the pull request to be automatically mergable :)

**Comment by robinfhu** [comment]
Can you also please remove any files from the build/ directory from this PR? Thanks!

**Comment by mgalloue** [comment]
@liquidpele @robinfhu , 
Sorry for all my bad commits I'm not very comfortable with GIT. I will try to do better the next time..
Regarding the files changed, I think it's OK now.

Thanks,

File: src/css/parallelcoordinates.css [hunk 6]
```diff
@@ -10,18 +10,25 @@
...
 
-.nvd3 .axis line, .axis path {
+.nvd3 .nv-axis line, .nv-axis path {
     fill: none;
     stroke: #000;
     shape-rendering: crispEdges;
 }
 
-.nvd3 .axis text {
+.nvd3 .nv-axis text {
```
Comment (liquidpele) [inline src/css/parallelcoordinates.css]: Why the change here?  This will apply to more than just this chart... were these even valid before?

File: src/models/parallelCoordinates.js [hunk 7]
```diff
@@ -1,7 +1,3 @@
-
-// Code adapted from Jason Davies' "Parallel Coordinates"
-// http://bl.ocks.org/jasondavies/1341281
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: Don't remove that please, it's nice to see the origins of things.

File: src/models/parallelCoordinates.js [hunk 8]
```diff
@@ -9,16 +5,18 @@ nv.models.parallelCoordinates = function() {
...
     //------------------------------------------------------------
 
-    var margin = {top: 30, right: 10, bottom: 10, left: 10}
+    var margin = {top: 30, right: 0, bottom: 10, left: 0}
         , width = null
         , height = null
         , x = d3.scale.ordinal()
         , y = {}
-        , dimensions = []
+        , dimensionsName = []
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: Should be "dimensionNames" if it's a list of things.

Comment (liquidpele) [inline src/models/parallelCoordinates.js]: Also you'll want to add a description of it in examples/documentation.html with the other option documentation (see the json at the end of the file)

File: src/models/parallelCoordinates.js [hunk 9]
```diff
@@ -157,14 +220,15 @@ nv.models.parallelCoordinates = function() {
         // simple options, just get/set the necessary values
         width:      {get: function(){return width;}, set: function(_){width=_;}},
         height:     {get: function(){return height;}, set: function(_){height=_;}},
-        dimensions: {get: function(){return dimensions;}, set: function(_){dimensions=_;}},
+        dimensionsName: { get: function() { return dimensionsName; }, set: function(_) { dimensionsName = _; } },
+        lineTension: {get: function(){return lineTension;}, set: function(_){lineTension=_;}},
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: Add documentation for lineTension in examples/documentation.html

File: src/models/parallelCoordinates.js [hunk 10]
```diff
@@ -157,14 +220,15 @@ nv.models.parallelCoordinates = function() {
         // simple options, just get/set the necessary values
         width:      {get: function(){return width;}, set: function(_){width=_;}},
         height:     {get: function(){return height;}, set: function(_){height=_;}},
-        dimensions: {get: function(){return dimensions;}, set: function(_){dimensions=_;}},
+        dimensionsName: { get: function() { return dimensionsName; }, set: function(_) { dimensionsName = _; } },
```
Comment (liquidpele) [inline src/models/parallelCoordinates.js]: Since your'e renaming the option, keep the dimension option there but log it to be deprecated.  See the tooltip and tooltipContent options in other charts for examples.

File: src/models/parallelCoordinates.js [hunk 11]
```diff
@@ -141,6 +171,39 @@ nv.models.parallelCoordinates = function() {
...
+                dragging[d] = this.parentNode.__origin__ = x(d);
+                background.attr("visibility", "hidden");
+
+            }
+
+            function dragMove(d, i) {
+                dragging[d] = Math.min(width, Math.max(0, this.parentNode.__origin__ += d3.event.x));
+                foreground.attr("d", path);
+                dimensionsName.sort(function(a, b) { return position(a) - position(b); });
+                //dimensions.map(function(d, i) { return d.currentPosition = i; });
```
Comment (robinfhu) [inline src/models/parallelCoordinates.js]: remove commented code

File: src/models/parallelCoordinates.js [hunk 12]
```diff
@@ -141,6 +171,39 @@ nv.models.parallelCoordinates = function() {
...
+                delete this.parentNode.__origin__;
+                delete dragging[d];
+                d3.select(this.parentNode).attr("transform", "translate(" + x(d) + ")");
+                foreground
+                  .attr("d", path);
+                background
+                  .attr("d", path)
+                  .attr("visibility", null);
+
+                //dispatch.dimensionsOrder(dimensionsName);
```
Comment (robinfhu) [inline src/models/parallelCoordinates.js]: remove commented code

File: src/models/parallelCoordinates.js (hunk 9 above)
Comment (mgalloue) [inline src/models/parallelCoordinates.js]: lineTension is already added in documentation.

File: src/models/parallelCoordinates.js (hunk 12 above)
Comment (mgalloue) [inline src/models/parallelCoordinates.js]: done

File: src/models/parallelCoordinates.js (hunk 11 above)
Comment (mgalloue) [inline src/models/parallelCoordinates.js]: done

File: src/models/parallelCoordinates.js (hunk 10 above)
Comment (mgalloue) [inline src/models/parallelCoordinates.js]: Done

File: src/css/parallelcoordinates.css (hunk 6 above)
Comment (mgalloue) [inline src/css/parallelcoordinates.css]: You're right, I removed it because it's not necessary.

File: src/models/parallelCoordinates.js (hunk 7 above)
Comment (mgalloue) [inline src/models/parallelCoordinates.js]: OK

**Comment by robinfhu** [comment]
Thanks!
